src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -868,17 +1039,222 @@ public bool MoveNext()
 
             public void Reset()
             {
-                _exactMatch = ExactMatchState.NotChecked;
-                _enumerator = default(Dictionary<string, ModelStateEntry>.Enumerator);
-                Current = default(KeyValuePair<string, ModelStateEntry>);
+                _index = 0;
+                _nodes.Clear();
+                _visitedRoot = false;
+                _modelStateNode = null;
             }
+        }
+
+        public struct KeyCollection : ICollection<string>
+        {
+            private readonly ModelStateDictionary _dictionary;
 
-            private enum ExactMatchState
+            public KeyCollection(ModelStateDictionary dictionary)
+            {
+                _dictionary = dictionary;
+            }
+
+            public int Count => _dictionary.Count;
+
+            public bool IsReadOnly => true;
+
+            public void Add(string item)
+            {
+                throw new NotSupportedException();
+            }
+
+            public void Clear()
+            {
+                throw new NotSupportedException();
+            }
+
+            public bool Contains(string item)
+            {
+                if (item == null)
+                {
+                    return false;
+                }
+
+                return _dictionary.ContainsKey(item);
+            }
+
+            public void CopyTo(string[] array, int arrayIndex)
+            {
+                if (array == null)
+                {
+                    throw new ArgumentNullException(nameof(array));
+                }
+
+                if (arrayIndex < 0 || arrayIndex + Count > array.Length)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(arrayIndex));
+                }
+
+                foreach (var item in this)
+                {
+                    array[arrayIndex++] = item;
+                }
+            }
+
+            public bool Remove(string item)
+            {
+                throw new NotSupportedException();
+            }
+
+            public KeyEnumerator GetEnumerator() => new KeyEnumerator(_dictionary, prefix: string.Empty);
+
+            IEnumerator<string> IEnumerable<string>.GetEnumerator() => GetEnumerator();
+
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+        }
+
+        public struct KeyEnumerator : IEnumerator<string>
```

Comment:
Two levels of nesting seems like overkill. Plus the type is perfectly usable outside of the enumerable.